lib/segment/src/index/hnsw_index/links_container.rs

```diff
@@ -0,0 +1,434 @@
+        } else if id_to_insert != self.links.len() {
+            self.links.pop();
+            self.links.insert(id_to_insert, new_point_id);
+        }
+    }
+
+    /// Append one point to the container. If the container is full, run the heuristic.
+    ///
+    /// This is a reference implementation for testing.
+    #[cfg(test)]
+    fn connect_with_heuristic_simple(
+        &mut self,
+        new_point_id: PointOffsetType,
+        target_point_id: PointOffsetType,
+        level_m: usize,
+        mut score: impl FnMut(PointOffsetType, PointOffsetType) -> ScoreType,
+    ) {
+        if self.links.len() < level_m {
+            self.links.push(new_point_id);
+        } else {
+            let mut candidates = Vec::with_capacity(level_m + 1);
+            for &idx in &self.links {
+                candidates.push(ScoredPointOffset {
+                    idx,
+                    score: score(target_point_id, idx),
+                });
+            }
+            candidates.push(ScoredPointOffset {
+                idx: new_point_id,
+                score: score(target_point_id, new_point_id),
+            });
+            candidates.sort_unstable_by(|a, b| b.score.total_cmp(&a.score));
+            self.fill_from_sorted_with_heuristic(candidates.into_iter(), level_m, score);
+        }
+    }
+
+    /// Append one point to the container. If the container is full, run the heuristic.
+    ///
+    /// The result is exactly the same as [`Self::connect_with_heuristic_simple`],
+    /// but this implementation cuts some corners given that some of the links
+    /// are already processed by the heuristic.
+    pub fn connect_with_heuristic(
+        &mut self,
+        new_point_id: PointOffsetType,
+        target_point_id: PointOffsetType,
+        level_m: usize,
+        mut score: impl FnMut(PointOffsetType, PointOffsetType) -> ScoreType,
+        items: &mut ItemsBuffer,
+    ) {
+        if level_m == 0 {
+            // Unlikely.
+            return;
+        }
+
+        if self.links.len() < level_m {
+            self.links.push(new_point_id);
+            return;
+        }
+
+        items.0.clear();
+        items.0.try_set_capacity_exact(level_m + 1).unwrap();
+        for (order, &link) in self.links.iter().enumerate() {
+            items.0.push(Item {
+                idx: link,
+                score: Cell::new(None),
+                order: if order < self.processed_by_heuristic as usize {
+                    NonZeroU32::new(order as u32)
+                } else {
+                    None
+                },
+            });
+        }
+        items.0.push(Item {
+            idx: new_point_id,
+            score: Cell::new(None),
+            order: None,
+        });
+        items.0.sort_unstable_by(|a, b| {
+            if let (Some(a_order), Some(b_order)) = (a.order, b.order) {
+                return a_order.cmp(&b_order);
+            }
+            b.score(target_point_id, &mut score)
```

Comment:
Do we assume that re-computing scores for already inserted points once is negligible?

Reply:
No, but I decided to not to preserve because it would impose 2x RAM increase (see "Further improvements" point 1 in the PR description). If we decide to save it, I suggest to have it in a separate PR.

Also, `.score()` is lazy/cached, the second call would return the saved result.

And calling `.score()` is skipped if both `a` and `b` are already processed by heuristics.